File: HOPPING.cpp
#include <complex>
#include <iostream>
#include <fstream>
#include <iomanip>
#include <vector>
#include <nlohmann/json.hpp> //we use json format for input file

using json = nlohmann::json;


template<typename T>
std::vector <size_t> sort_indexes(const std::vector <T> &v) {

    std::vector <size_t> idx(v.size());
    std::iota(idx.begin(), idx.end(), 0);
    std::stable_sort(idx.begin(), idx.end(),
                     [&v](size_t i1, size_t i2) { return v[i1] < v[i2]; });

    return idx;
}

void
coordination_sort(int atom, int num_atoms, int n_min[3], int n_max[3], double cell_vectors[3][3],
                  std::vector <std::vector<double>> &positions,
                  std::vector<double> &radius, std::vector <std::vector<int>> &index) {


    int i, j, k, x, z;
    int num_points;

    int n_size[3];
    double r[3];


    for (i = 0; i < 3; i++) {
        n_size[i] = n_max[i] - n_min[i] + 1; // Plus 1 for 0th
    }

    num_points = num_atoms * (n_size[0] * n_size[1] * n_size[2]);


    // index = [i, j, k, atom]
    std::vector <std::vector<int>> idx(
            num_points, std::vector<int>(
                    4)
    );


    std::vector<double> rad(num_points);


    // Index filling
    i = n_min[0];
    j = n_min[1];
    k = n_min[2];
    x = 0;
    for (z = 0; z < num_points; z++) {
        idx[z][0] = i;
        idx[z][1] = j;
        idx[z][2] = k;
        idx[z][3] = x;
        r[0] = i * cell_vectors[0][0] + j * cell_vectors[1][0] + k * cell_vectors[2][0] +
               (positions[x][0] - positions[atom][0]);
        r[1] = i * cell_vectors[0][1] + j * cell_vectors[1][1] + k * cell_vectors[2][1] +
               (positions[x][1] - positions[atom][1]);
        r[2] = i * cell_vectors[0][2] + j * cell_vectors[1][2] + k * cell_vectors[2][2] +
               (positions[x][2] - positions[atom][2]);
        rad[z] = sqrt(r[0] * r[0] + r[1] * r[1] + r[2] * r[2]);
        x++;
        if (x == num_atoms) {
            x = 0;
            k++;
            if (k == n_max[2] + 1) {
                k = n_min[2];
                j++;
            }
            if (j == n_max[1] + 1) {
                j = n_min[1];
                i++;
            }

        }
    }


    k = 0;
    for (auto i: sort_indexes(rad)) {
        for (j = 0; j < 4; j++) {
            index[k][j] = idx[i][j];
        }
        radius[k] = rad[i];
        k++;

    }


}


int main() {

    time_t td;
    int i, j, x, y, z;
    int n_size[3], n_min[3], n_max[3], vecs[3], orbs[2];
    int num_atoms, num_wannier_funcs, num_points, max_sphere_num, neighbor_num, sphere_num, move_x, move_y, move_x1, move_y1, print_complex, atom;

    double cell_vectors[3][3], ham_values[2], r[3];

    std::string input_file_name;


    td = time(NULL);

    std::cout << "Program HOPPING.x v.3.0 (c++) starts on ";
    std::cout << ctime(&td) << std::endl;
    std::cout << "=====================================================================" << std::endl;

    std::cout << "Enter the name of input file (name.json and name_hr.dat)" << std::endl;
    std::cin >> input_file_name;


    std::ifstream hr;
    hr.open(input_file_name + "_hr.dat");
    if (!hr) {
        std::cout << "ERROR! Cannot open hopping file " << input_file_name << "_hr.dat" << std::endl;
        return 0;
    }



    //define the possible ranges for n_min and n_max from hr.dat file
    for (i = 0; i < 3; i++) {
        n_max[i] = n_min[i] = 0;
    }

    while (!hr.eof()) {
        hr >> vecs[0] >> vecs[1] >> vecs[2] >> orbs[0] >> orbs[1] >> ham_values[0] >> ham_values[1];

        for (i = 0; i < 3; i++) {
            if (n_max[i] < vecs[i]) n_max[i] = vecs[i];
            else if (n_min[i] > vecs[i]) n_min[i] = vecs[i];
        }

    }


    for (i = 0; i < 3; i++) {
        n_size[i] = n_max[i] - n_min[i] + 1; // Plus 1 for 0th
    }


    //===============================================================================
    //read information from input file in json format

    std::ifstream in;
    in.open(input_file_name + ".json");
    if (!in) {
        std::cout << "ERROR! Cannot open input file " << input_file_name << ".json" << std::endl;
        return 0;
    }


    json json_file;
    in >> json_file;

    in.close();


    // read int numbers from json file
    json_file.at("number_of_atoms").get_to(num_atoms);
    json_file.at("max_sphere_num").get_to(max_sphere_num);
    json_file.at("print_complex").get_to(print_complex);


    std::vector <std::vector<double>> positions(
            num_atoms, std::vector<double>(
                    3)
    );


    std::vector<double> wanniers(num_atoms);


    // read matrices from json file

    i = 0;
    for (auto &element : json_file["wanniers"]) {
        wanniers[i] = element;
        i++;
    }


    i = 0;
    for (auto &element : json_file["cell_vectors"]) {
        j = 0;
        for (auto &element1 : element) {
            cell_vectors[i][j] = element1;
            j++;
        }
        i++;
    }


    i = 0;
    for (auto &element : json_file["positions"]) {
        j = 0;
        for (auto &element1 : element) {
            positions[i][j] = element1;
            j++;
        }
        i++;
    }


    //total number of Wannier functions
    num_wannier_funcs = std::accumulate(wanniers.begin(), wanniers.end(), 0);


    // Total hamiltonian
    std::vector < std::vector < std::vector < std::vector < std::vector < std::complex<double> >> >> > Ham_R(
            n_size[0], std::vector < std::vector < std::vector < std::vector < std::complex<double> >> >> (
                    n_size[1], std::vector < std::vector < std::vector < std::complex<double> >> > (
                            n_size[2], std::vector < std::vector < std::complex<double> >> (
                                    num_wannier_funcs, std::vector< std::complex<double> >(
                                            num_wannier_funcs)
                            )
                    )
            )
    );


    hr.clear();
    hr.seekg(0, std::ios::beg);


    while (!hr.eof()) {
        hr >> vecs[0] >> vecs[1] >> vecs[2] >> orbs[0] >> orbs[1] >> ham_values[0] >> ham_values[1];

        Ham_R[vecs[0] + n_max[0]][vecs[1] + n_max[1]][vecs[2] + n_max[2]][orbs[0] - 1][orbs[1] -
                                                                                       1] = std::complex<double>(
                ham_values[0], ham_values[1]);
    }
    std::cout << "Hopping file " << input_file_name << "_hr.dat was  scanned  successfully" << std::endl;

    hr.close();


    //===============================================================================
    // print data to stdout and out-file


    std::ofstream out;
    out.open("HOPPING.dat");


    std::cout << "Crystal structure of system:" << std::endl;
    out << "Crystal structure of system:" << std::endl;

    std::cout << "crystal axes: (cart. coord. )" << std::endl;
    out << "crystal axes: (cart. coord. )" << std::endl;


    for (i = 0; i < 3; i++) {

        std::cout << std::fixed << std::setprecision(6) << cell_vectors[i][0] << "  " << cell_vectors[i][1] << "  "
                  << cell_vectors[i][2]
                  << std::endl;
        out << std::fixed << std::setprecision(6) << cell_vectors[i][0] << "  " << cell_vectors[i][1] << "  "
            << cell_vectors[i][2]
            << std::endl;
    }

    std::cout << std::endl;
    out << std::endl;

    std::cout << "Number of atoms: " << num_atoms << std::endl;
    out << "Number of atoms: " << num_atoms << std::endl;

    std::cout << std::endl << "Atomic positions (cart. coord.):\n x\t y\t z\t number of WF:" << std::endl;
    out << std::endl << "Atomic positions (cart. coord. ):\n x\t y\t z\t number of WF:" << std::endl;


    for (i = 0; i < num_atoms; i++) {
        std::cout << std::fixed << std::setprecision(6) << positions[i][0] << "  " << positions[i][1] << "  "
                  << positions[i][2] << "  "
                  << wanniers[i]
                  << std::endl;
        out << std::fixed << std::setprecision(6) << positions[i][0] << "  " << positions[i][1] << "  "
            << positions[i][2]
            << "  "
            << wanniers[i]
            << std::endl;
    }

    std::cout << "Total number of Wannier functions: " << num_wannier_funcs << std::endl;
    out << "Total number of Wannier functions: " << num_wannier_funcs << std::endl;


    std::cout << "Print the complex part of the hamiltonian: " << print_complex << std::endl;
    out << "Print the complex part of the hamiltonian: " << print_complex << std::endl;


    std::cout << std::endl;
    out << std::endl;


    //==============================================================================

    // move_x and move_y control the shift to the necessary block in Ham_R
    move_x = move_y = 0;

    num_points = num_atoms * (n_size[0] * n_size[1] * n_size[2]);


    // index = [i, j, k, atom]
    std::vector <std::vector<int>> index(
            num_points, std::vector<int>(
                    4)
    );


    std::vector<double> radius(num_points);



    //'atom' is the index of the central atom
    for (atom = 0; atom < num_atoms; atom++) {


        //This function sorts atoms depending on the radius from the central atom with index 'atom'
        coordination_sort(atom, num_atoms, n_min, n_max, cell_vectors, positions, radius, index);


        out << "---------------------------------------------------------------------" << std::endl;

        neighbor_num = 1;
        sphere_num = 0;


        for (z = 0; z < num_points; z++) {

            if (z != 0) {
                if (fabs(radius[z - 1] - radius[z]) < 1E-4)neighbor_num++;
                else {
                    neighbor_num = 1;
                    sphere_num++;
                }
            }
            if (sphere_num == max_sphere_num) break;


            out << "Atom " << atom + 1 << "(000)<-->Atom " << index[z][3] + 1 << "(" << index[z][0] << index[z][1]
                << index[z][2] << ") in sphere #" << sphere_num << " with radius " << radius[z] << " is #"
                << neighbor_num
                << ":  " << std::endl;


            r[0] = index[z][0] * cell_vectors[0][0] + index[z][1] * cell_vectors[1][0] +
                   index[z][2] * cell_vectors[2][0] +
                   (positions[index[z][3]][0] - positions[atom][0]);
            r[1] = index[z][0] * cell_vectors[0][1] + index[z][1] * cell_vectors[1][1] +
                   index[z][2] * cell_vectors[2][1] +
                   (positions[index[z][3]][1] - positions[atom][1]);
            r[2] = index[z][0] * cell_vectors[0][2] + index[z][1] * cell_vectors[1][2] +
                   index[z][2] * cell_vectors[2][2] +
                   (positions[index[z][3]][2] - positions[atom][2]);

            out << std::fixed << std::setprecision(6) << "Radius vector is: " << r[0] << " " << r[1] << " " << r[2]
                << std::endl;

            move_y = 0;
            for (x = 0; x < index[z][3]; x++) {
                move_y += wanniers[x];
            }

            move_x1 = wanniers[atom] + move_x;
            move_y1 = wanniers[index[z][3]] + move_y;


            if (print_complex == 1) {
                for (x = move_x; x < move_x1; x++) {
                    for (y = move_y; y < move_y1; y++) {


                        out << std::fixed << std::setprecision(6)
                            << Ham_R[index[z][0] + n_max[0]][index[z][1] + n_max[1]][index[z][2] + n_max[2]][x][y]
                            << "  ";

                    }
                    out << std::endl;
                }
                out << std::endl;
            } else {
                for (x = move_x; x < move_x1; x++) {
                    for (y = move_y; y < move_y1; y++) {


                        out << std::fixed << std::setprecision(6)
                            << (Ham_R[index[z][0] + n_max[0]][index[z][1] + n_max[1]][index[z][2] +
                                                                                      n_max[2]][x][y]).real()
                            << "  ";

                    }
                    out << std::endl;
                }
                out << std::endl;
            }

        }
        move_x += wanniers[atom];


    }


    td = time(NULL);
    std::cout << std::endl << "This run was terminated on: " << std::endl;
    std::cout << ctime(&td) << std::endl;

    std::cout << "JOB DONE" << std::endl;
    std::cout << "=====================================================================" << std::endl;


}

File: HOPPING.py
import numpy as np
import json
from io import StringIO
from typing import OrderedDict
from datetime import datetime
import argparse

#https://triqs.github.io/tprf/latest/reference/python_reference.html#wannier90-tight-binding-parsers
def parse_hopping_from_wannier90_hr_dat(filename):

    with open(filename, 'r') as fd:
        lines = fd.readlines()

    lines.pop(0) # pop time header

    num_wann = int(lines.pop(0))
    nrpts = int(lines.pop(0))

    nlines = int(np.ceil(float(nrpts / 15.)))

    deg = np.array([])
    for line in lines[:nlines]:
        deg = np.concatenate((deg, np.loadtxt(StringIO(line), dtype=int, ndmin=1)))
    
    assert( deg.shape == (nrpts,) )

    hopp = "".join(lines[nlines:])
    hopp = np.loadtxt(StringIO(hopp))

    assert( hopp.shape == (num_wann**2 * nrpts, 7) )
    
    R = np.array(hopp[:, :3], dtype=int) # Lattice coordinates in multiples of lattice vectors
    nm = np.array(hopp[:, 3:5], dtype=int) - 1 # orbital index pairs, wannier90 counts from 1, fix by remove 1
    
    n_min, n_max = R.min(0), R.max(0)

    t_re = hopp[:, 5]
    t_im = hopp[:, 6]
    t = t_re + 1.j * t_im # complex hopping amplitudes for each R, mn (H(R)_{mn})

    # -- Dict with hopping matrices

    r_dict = OrderedDict()
    hopp_dict = {}
    for idx in range(R.shape[0]):
        r = tuple(R[idx])

        if r not in r_dict:
            r_dict[r] = 1
        else:
            r_dict[r] += 1

        if r not in hopp_dict:
            hopp_dict[r] = np.zeros((num_wann, num_wann), dtype=complex)

        n, m = nm[idx]
        hopp_dict[r][n, m] = t[idx]

    # -- Account for degeneracy of the Wigner-Seitz points
    
    for r, weight in zip(list(r_dict.keys()), deg):
        hopp_dict[r] /= weight

    return hopp_dict, num_wann, n_min, n_max



def coordination_sort(atom, num_atoms, n_min, n_max, cell_vectors, positions):

    n_size = n_max - n_min + 1  # Plus 1 for 0th
    num_points = num_atoms * n_size.prod()

    # Index filling
    index_ = np.stack(np.unravel_index(np.arange(num_points),
                                       (*n_size, num_atoms)),
                      axis=-1)
    index_ += [*n_min, 0]

    r = index_[:, :3] @ cell_vectors + positions[index_[:, 3]] - positions[atom]
    radius_ = (r * r).sum(-1)**0.5

    # Sort radius descending
    idx = radius_.argsort()
    return radius_[idx], index_[idx]


def main():
    """
    This script converts input_hr.dat hamiltonian from the wannier90 package to block-divided 
    form between orbitals and prints these blocks depending on the geometry given by input.json file.

    Usage: python HOPPING.py input out.dat

    input.json contains the following information:
    "cell_vectors": (3x3)(dfloat) array of unit cell vectors (in Ang);
    "number_of_atoms": (int) total number of atoms, parametrized by Wannier functions;
    "max_sphere_num": (int) maximum number of coordination sphere to print the hamiltonian blocks between pair of atoms;
    "print_complex": (0 or 1) print the imaginary part of hamiltonian blocks;
    "positions": (3 x number_of_atoms)(dfloat) array of cartesian coordinates of these atoms (in Ang);
    "wanniers":  (number_of_magnetic_atoms) (int) array of Wannier functions numbers of these atoms;
    """
    
    parser = argparse.ArgumentParser(prog='HOPPING.py', usage='%(prog)s input (name of input_hr.dat and input.json) output')
    parser.add_argument("input_file_name")
    parser.add_argument("output_file_name")
    args = parser.parse_args()


    print("Program HOPPING.x v.3.0 (python) starts on ", datetime.now())
    print('=' * 69)

    hops, num_wannier_funcs, n_min, n_max = parse_hopping_from_wannier90_hr_dat(f'{args.input_file_name}_hr.dat') 
    n_size = n_max - n_min + 1  # Plus 1 for 0th

    Ham_R = np.zeros((n_size[0], n_size[1], n_size[2], num_wannier_funcs, num_wannier_funcs), dtype='c16')
    
    for r in hops.keys():
        r_idx = np.array(r)
        Ham_idx = hops.get(r)
        
        for m in range(num_wannier_funcs):
            for n in range(num_wannier_funcs):
                Ham_R[r_idx[0] + n_max[0], r_idx[1] + n_max[1], r_idx[2] + n_max[2], m, n] = Ham_idx[m,n]

    

    # =======================================================================
    # Read information from input file in json format
    with open(f'{args.input_file_name}.json') as fp:
        data = json.load(fp)

    num_atoms = data['number_of_atoms']  # M
    positions = np.array(data['positions'])  # array[M, 3]
    wanniers = np.array(data['wanniers'])  # array[M]
    cell_vectors = np.array(data['cell_vectors'])  # array[3, 3]
    max_sphere_num = data['max_sphere_num']
    print_complex = data['print_complex']  # Pretty repr

    # Some checks
    assert num_atoms == len(wanniers) == len(positions)
    assert num_wannier_funcs == wanniers.sum()

    # ===============================================================================
    # Print data to stdout and out-file

    print('Crystal structure of system:')
    print('crystal axes: (cart. coord. )')

    for cell in cell_vectors:
        print('{:.6f} {:.6f} {:.6f}'.format(*cell))

    print(f'\nNumber of atoms: {num_atoms}')
    print('\nAtomic positions (cart. coord.):' '\n x\t y\t z\t number of WF:')
    for pos, wan in zip(positions, wanniers):
        print('{:.6f} {:.6f} {:.6f} {:.6f}'.format(*pos, wan))

    print(
        f'\nTotal number of Wannier functions: {num_wannier_funcs}'
        f'\nPrint the complex part of the hamiltonian: {bool(print_complex)}')
    with open(args.output_file_name, 'w') as fp:
        print('Crystal structure of system:', file=fp)
        print('crystal axes: (cart. coord. )', file=fp)
        for cell in cell_vectors:
            print('{:.6f} {:.6f} {:.6f}'.format(*cell), file=fp)
        print('\nNumber of atoms: ', num_atoms, file=fp)
        print(
            '\nAtomic positions (cart. coord.):'
            '\n x\t y\t z\t number of WF:',
            file=fp)
        for pos, wan in zip(positions, wanniers):
            print('{:.6f} {:.6f} {:.6f} {:.6f}'.format(*pos, wan), file=fp)
        print('\nTotal number of Wannier functions: ',
              num_wannier_funcs,
              file=fp)
        print('\nPrint the complex part of the hamiltonian: ',
              print_complex,
              file=fp)
        
    # =======================================================================

    # move_x and move_y control the shift to the necessary block in Ham_R
    move_x = move_y = 0

    num_points = num_atoms * n_size.prod()
    index = np.zeros((num_points, 4), dtype=int)  # index = [i, j, k, atom]
    radius = np.zeros(num_points, dtype=float)

    # 'atom' is the index of the central atom
    with open(args.output_file_name, 'a') as fp:
        for atom, (pos, wan) in enumerate(zip(positions, wanniers)):
            # This function sorts atoms depending on the radius from the central atom with index 'atom'
            radius, index = coordination_sort(atom, num_atoms, n_min, n_max, cell_vectors,
                                              positions)

            print('-' * 69, file=fp)

            neighbor_num = 1
            sphere_num = 0
            for z, (rad, idx) in enumerate(zip(radius, index)):
                if z:
                    if abs(radius[z - 1] - rad) < 1e-4:
                        neighbor_num += 1
                    else:
                        neighbor_num = 1
                        sphere_num += 1

                if sphere_num == max_sphere_num:
                    break

                d3, at = idx[:3], idx[3]
                r = d3 @ cell_vectors + (positions[at] - pos)

                print(f'Hopping <a|H|b> between atom  {atom} (000)<-->atom  {at}',
                      '(',
                      *d3,
                      f') in sphere # {sphere_num}  '
                      'with radius  {:.6f}'.format(rad),
                      f' -- {neighbor_num}:',
                      file=fp)
                print('Radius vector is:  {:.6f} {:.6f} {:.6f}'.format(*r),
                      file=fp)

                move_y = wanniers[:at].sum()
                s = Ham_R[tuple(d3 + n_max)][move_x:,
                    move_y:][:wan, :wanniers[at]]

                fmt = '({0.real:.4f}  {0.imag:.4f}i)' if print_complex else '{0.real:.6f}'
                print('\n'.join('  '.join(fmt.format(item) for item in row)
                                for row in s),
                      file=fp)
                print(file=fp)

            move_x += wan

    print(f'This run was terminated on: {datetime.now()}')
    print(f'JOB DONE')
    print('=' * 69)



if __name__ == '__main__':
    main()

    
